packages/flutter/lib/src/material/input.dart

```diff
@@ -13,6 +13,126 @@ import 'theme.dart';
 
 export 'package:sky_services/editing/editing.mojom.dart' show KeyboardType;
 
+/// A container for grouping together multiple Input widgets.
+class Form extends StatefulWidget {
+  Form({
+    Key key,
+    this.child,
+    this.data,
+    this.onSubmitted
+  }) : super(key: key) {
+    assert(child != null);
+  }
+
+  /// A map of form field IDs to their input values.
+  final Map<String, String> data;
+
+  /// Called when the input is accepted anywhere on the form.
+  final VoidCallback onSubmitted;
+
+  /// Root of the widget hierarchy that contains this form.
+  final Widget child;
+
+  @override
+  _FormState createState() => new _FormState();
+}
+
+class _FormState extends State<Form> {
+  // A map of form field IDs to their state.
+  final Map<String, _FormFieldState> fieldState = <String, _FormFieldState>{};
+
+  void onChanged(String id, InputValue value) {
+    config.data[id] = value.text;
+    // We sometimes need other inputs to respond to changes (e.g. a password
+    // verifier should update its error text if the other password field
+    // changes).
+    fieldState.forEach((String key, _FormFieldState data) {
+      data.inputState.setState((){ });
```

Comment:
always put a comment in the block of a setState saying what state it is you are changing.